lib/rpmsg/rpmsg.c

```diff
@@ -316,38 +304,113 @@ int rpmsg_init_vdev(struct rpmsg_virtio_device *rvdev,
 		vq->shm_io = shm_io;
 	}
 
-	if (rpmsg_virtio_get_role(rvdev) == RPMSG_MASTER) {
-		struct virtqueue_buf vqbuf;
-		unsigned int idx;
-		void * buffer;
+	vqbuf.len = RPMSG_BUFFER_SIZE;
+	for (idx = 0; ((idx < rvdev->rvq->vq_nentries)
+			&& (idx < rvdev->shbuf->total_buffs / 2));
+		idx++) {
+
+		/* Initialize TX virtqueue buffers for remote device */
+		buffer = sh_mem_get_buffer(rvdev->shbuf);
+
+		if (!buffer) {
+			return RPMSG_ERR_NO_BUFF;
+		}
+
+		vqbuf.buf = buffer;
+
+		metal_io_block_set(shm_io,
+			metal_io_virt_to_offset(shm_io, buffer),
+			0x00,
+			RPMSG_BUFFER_SIZE);
+		status =
+			virtqueue_add_buffer(rvdev->rvq, &vqbuf, 0, 1,
+						buffer);
+
+		if (status != RPMSG_SUCCESS) {
+			return status;
+		}
+	}
+
+	/* Initialize channels and endpoints list */
+	metal_list_init(&rvdev->endpoints);
+
+	dev_features = rpmsg_virtio_get_features(rvdev);
+
+	/*
+	 * Create name service announcement endpoint if device supports name
+	 * service announcement feature.
+	 */
+	if ((dev_features & (1 << VIRTIO_RPMSG_F_NS))) {
+		ns_ept.addr = RPMSG_NS_EPT_ADDR;
+		ns_ept.cb = rpmsg_ns_callback;
+		strncpy(ns_ept.name, "NS", sizeof("NS"));
+		status = rpmsg_register_endpoint(rvdev, &ns_ept);
+
+	}
```

Comment:
check virtio features and endpoint creation is the common, i think can have a common function for them

Reply:
i will have a look to this,